prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,292 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <config.h>
+#include <stdlib.h>
+#include <fi_util.h>
+#include <fi_iov.h>
+#include <ofi_mr.h>
+#include <fi_list.h>
+
+
+static int util_mr_find_overlap(void *a, void *b)
+{
+	struct iovec *iov1 = a, *iov2 = b;
+
+	if (ofi_iov_left(iov1, iov2))
+		return -1;
+	else if (ofi_iov_right(iov1, iov2))
+		return 1;
+	else
+		return 0;
+}
+
+static void util_mr_free_entry(struct ofi_mr_cache *cache,
+			       struct ofi_mr_entry *entry)
+{
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "free %p (len: %" PRIu64 ")\n",
+	       entry->iov.iov_base, entry->iov.iov_len);
+
+	assert(!entry->cached);
+	if (entry->subscribed) {
+		ofi_monitor_unsubscribe(entry->iov.iov_base, entry->iov.iov_len,
+					&entry->subscription);
+	}
+	cache->delete_region(cache, entry);
+	free(entry);
+	cache->cached_cnt--;
+}
+
+static void util_mr_uncache_entry(struct ofi_mr_cache *cache,
+				  struct ofi_mr_entry *entry)
+{
+	RbtIterator iter;
+
+	assert(entry->cached);
+	iter = rbtFind(cache->mr_tree, &entry->iov);
+	assert(iter);
+	rbtErase(cache->mr_tree, iter);
+	entry->cached = 0;
+}
+
+static void
+util_mr_cache_process_events(struct ofi_mr_cache *cache)
+{
+	struct ofi_subscription *subscription;
+	struct ofi_mr_entry *entry;
+
+	while ((subscription = ofi_monitor_get_event(&cache->nq))) {
+		entry = container_of(subscription, struct ofi_mr_entry,
+				     subscription);
+		if (entry->cached)
+			util_mr_uncache_entry(cache, entry);
+
+		if (entry->use_cnt == 0) {
+			dlist_remove(&entry->lru_entry);
+			util_mr_free_entry(cache, entry);
+		}
+	}
+}
+
+bool ofi_mr_cache_flush(struct ofi_mr_cache *cache)
+{
+	struct ofi_mr_entry *entry;
+
+	if (dlist_empty(&cache->lru_list))
+		return false;
+
+	dlist_pop_front(&cache->lru_list, struct ofi_mr_entry,
+			entry, lru_entry);
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "flush %p (len: %" PRIu64 ")\n",
+	       entry->iov.iov_base, entry->iov.iov_len);
+
+	util_mr_uncache_entry(cache, entry);
+	util_mr_free_entry(cache, entry);
+	return true;
+}
+
+void ofi_mr_cache_delete(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
+{
+	FI_DBG(cache->domain->prov, FI_LOG_MR, "delete %p (len: %" PRIu64 ")\n",
+	       entry->iov.iov_base, entry->iov.iov_len);
+	cache->delete_cnt++;
+
+	util_mr_cache_process_events(cache);
+
```

Comment:
To handle the case where the cache is not required to support lazy deregistration, an additional flag should be added to the cache structure which should indicate whether the feature is in use. After that, the following should be added to the condition below.

```C 

     if (--entry->use_cnt == 0) {
		if (entry->cached && cache->lazy_deregistration) {
			dlist_insert_tail(&entry->lru_entry, &cache->lru_list);
		} else {
			util_mr_free_entry(cache, entry);
		}
	}

Reply:
Primarily, this is to support application cases where the monitor systems available via libfabric will not be sufficient, in which case the cache uses best-effort caching and aggressively purges entries on calls to 'delete'. 

The case that Dmitry is observing is a similar to the above.

Reply:
Setting the size = 0 will disable the cache.

Reply:
I guess the cache::size = 1 provides MR cache w/o lazy deregistration

Reply:
cache::size = 1 will still cache one entry, and has the potential to create the same problem -- it would just be incredibly unlikely.